File: 1_get_master_files.pl
###########################################
### get_master_files.R                  ###
### Author: Morad Elsaify               ###
### Date created: 02/06/20              ###
### Date modified: 03/26/20             ###
###########################################

###########################################################################################################
### This file downloads the master files from EDGAR. Master files contain a list of all filings by SEC  ###
### reporting institutions and list details of filing CIK, name, form type, and address. Perl code to   ###
### download master files copied almost verbatim from the LWP::UserAgent documentation                  ###
### (https://metacpan.org/pod/LWP::UserAgent). Note: Mozilla::CA is needed to get around credential     ###
### requirements that come with accessing https servers. If not already installed, can be installed via ###
### the terminal with the command sudo cpan Mozilla::CA (make sure command line tools are installed).   ###
###########################################################################################################

# use proper packages
use LWP::UserAgent;
use Mozilla::CA;
use Cwd;

# change working directory
chdir('/hps/group/fuqua/mie4/EDGAR Parsing/Data/Master Files') or die "cannot change: $!\n";

# misc setup
my $ua = LWP::UserAgent->new; 
$ua->timeout(10);
$ua->env_proxy;

# iterate over years
for($year = 1993; $year < 2021; $year = $year + 1) {
    # iterate over quarter
    for($i = 1; $i < 5; $i = $i + 1) {
        # make file address
        $quarter = 'QTR' . $i;
        $filegrag = 'https://www.sec.gov/Archives/edgar/full-index/' . $year . '/' . $quarter . '/master.idx';

        # get file from EDGAR
        my $response = $ua->get($filegrag);

        # pipe the output to a file named appropriately
        $filename = $year . $quarter . 'master';
        open(MYOUTFILE, '>' . $filename);
        if($response->is_success) {
            print MYOUTFILE $response->decoded_content;
        } else {
            die $response->status_line;
        }
        close(MYOUTFILE);
    }
}

File: 1_get_master_files.R
###########################################
### 1_get_master_files.R                ###
### Author: Morad Elsaify               ###
### Date created: 03/27/20              ###
### Date modified: 03/27/20             ###
###########################################

###########################################################################################################
### This file downloads the master files from EDGAR. Master files contain a list of all filings by SEC  ###
### reporting institutions and list details of filing CIK, name, form type, and address.                ###
###########################################################################################################

# source('/hps/group/fuqua/mie4/data_projects/edgar_parsing/code/1_get_master_files.R', echo = TRUE)

# clear
rm(list = ls())

# misc
options(width = 135) # set console width
options(menu.graphics = FALSE) # disable menu popups in XQuartz

# set seed
set.seed(05041995)

# load packages
library(data.table)
library(zoo)

# set directory
setwd('/hpc/group/fuqua/mie4/data_projects/edgar_parsing/data')

# set output folder
output_folder <- 'master_files'
if(!dir.exists(output_folder)) dir.create(output_folder)

# set start, end dates for master files
start_year <- 1993
end_year <- year(Sys.Date())
end_quarter <- ceiling(month(Sys.Date())/3)

##### DOWNLOAD THE MASTER FILES #####

# iterate over years
for(year in start_year:end_year) {
    # iterate over quarter
    for(quarter in paste0('QTR', 1:4)) {
        # get file address
        address <- paste('https://www.sec.gov/Archives/edgar/full-index', year, quarter, 'master.idx', sep = '/')

        # download the file
        download.file(address, paste(output_folder, paste0(year, quarter, 'master.txt'), sep = '/'))

        # if in end year and exceed end_quarter, break
        if(year == end_year & as.numeric(gsub('QTR', '', quarter)) >= end_quarter) break 
    }
}

##########
